### PR TITLE
Port the Alpakka sns sink

### PR DESCRIPTION
**Pull request by SeanFarrow**
This ports the sns graph stage to akka.net. the original scala implementation is at:
https://github.com/akka/alpakka/blob/master/sns/src/main/scala/akka/stream/alpakka/sns/SnsPublishFlowStage.scala

### CODE REVIEW

File: SNS/README.md [hunk 1]
```diff
@@ -0,0 +1,80 @@
+## Akka.Streams.SignalR
```
Comment (Silv3rcircl3) [inline SNS/README.md]: @SeanFarrow I think you've ported the wrong readme 😄

**Comment by SeanFarrow** [comment]
@Silv3rcircl3,
There wasn't a readme for this component, do you want me to add one? I left the SignalR one there as a reminder!

File: SNS/src/Akka.Streams.SNS/SNSPublishFlowStage.cs [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+                if (IsClosed(stage.inlet) && !isMessageInFlight) CompleteStage();
+                if (!HasBeenPulled(stage.inlet)) TryPull(stage.inlet);
+            }
+            
+            public override void OnUpstreamFinish()
+            {
+                if (!isMessageInFlight) CompleteStage();
+
+                SetHandler(stage.inlet, this);
+                SetHandler(stage.outlet, this);
```
Comment (Silv3rcircl3) [inline SNS/src/Akka.Streams.SNS/SNSPublishFlowStage.cs]: those two lines need to be moved in the constructor of the `Logic` class

File: SNS/src/Akka.Streams.SNS/SNSPublishFlowStage.cs [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+            }
+        }
+        #endregion
+
+        public override FlowShape<string, PublishResponse> Shape
+        {
+            get { return new FlowShape<string,PublishResponse>(inlet, outlet); }
```
Comment (Silv3rcircl3) [inline SNS/src/Akka.Streams.SNS/SNSPublishFlowStage.cs]: No need to create a new `Shape` on every call, I would move it into the constructor `Shape = new FlowShape<string,PublishResponse>(inlet, outlet);`

File: SNS/src/Akka.Streams.SNS/SNSPublisher.cs [hunk 4]
```diff
@@ -0,0 +1,13 @@
...
+using Amazon.SimpleNotificationService;
+using Amazon.SimpleNotificationService.Model;
+namespace Akka.Streams.SNS
+{
+    public static class SnsPublisher
+    {
+        public static Flow<string, PublishResponse, NotUsed> PublishToSNSFlow(string topicArn, IAmazonSimpleNotificationService snsService)
+        {
+            return Flow.FromGraph(new SNSPublishFlowStage(topicArn, snsService));
+        }
```
Comment (Silv3rcircl3) [inline SNS/src/Akka.Streams.SNS/SNSPublisher.cs]: Sink API is missing

File: SNS/src/Akka.Streams.SNS/Properties/AssemblyInfo.cs [hunk 5]
```diff
@@ -0,0 +1,36 @@
+﻿using System.Reflection;
+using System.Runtime.CompilerServices;
+using System.Runtime.InteropServices;
+
+// General Information about an assembly is controlled through the following 
+// set of attributes. Change these attribute values to modify the information
+// associated with an assembly.
+[assembly: AssemblyTitle("Akka.Streams.SignalR")]
```
Comment (Silv3rcircl3) [inline SNS/src/Akka.Streams.SNS/Properties/AssemblyInfo.cs]: `AssemblyTitle` and `AssemblyProduct` are referencing `SignalR`

File: SNS/src/Akka.Streams.SNS.Tests/Properties/AssemblyInfo.cs [hunk 6]
```diff
@@ -0,0 +1,36 @@
+﻿using System.Reflection;
+using System.Runtime.CompilerServices;
+using System.Runtime.InteropServices;
+
+// General Information about an assembly is controlled through the following 
+// set of attributes. Change these attribute values to modify the information
+// associated with an assembly.
+[assembly: AssemblyTitle("Akka.Streams.SignalR.Tests")]
```
Comment (Silv3rcircl3) [inline SNS/src/Akka.Streams.SNS.Tests/Properties/AssemblyInfo.cs]: `AssemblyTitle` and `AssemblyProduct` are referencing `SignalR`

**Comment by SeanFarrow** [comment]
What do you mean the sink api is missing?

**Comment by Silv3rcircl3** [comment]
Current API implementation only returns a `Flow<string, PublishedResult, NotUsed>` , i.e. `PublishToSNSFlow` but there should be another method which returns `Sink<string, Task>`, i.e. `PublishToSNSSink`

and regarding the readme, if you want to write a simple one go ahead otherwise remove it completely.

**Comment by SeanFarrow** [comment]
Ok, when would yo use ither then?

**Comment by Silv3rcircl3** [comment]
If you're not interested in the `PublishedResult` , it's not a big deal since you could write it yourself with a few method calls, but it's exists in the scala version so I mentioned it ;-)

**Comment by SeanFarrow** [comment]
Does this just need to be Task, or Task<PublishResult>, the scala is Future[Done].

**Comment by Silv3rcircl3** [comment]
Simple Task, Done is only used because they always need to returnsomethin in Scala

**Pull request merged**